src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -629,68 +637,53 @@ internal void Activate(object owner)
             Debug.Assert(result == 1, "invalid deactivate count");
         }
 
+        internal bool SetCancelStateClosed()
```

Comment:
Did you mean **SetCancelStateUnset**?

Reply:
No. CompareExchange is confusing, it changes the value in the ref for the first arg to the value in the second if the current value is the third, and then returns the original value not the new value. So this changes it from unset to closed and returns a boolean which identifies if it succeeded in changing to closed.

Reply:
You were wrong about the name but right that this code wasn't correct.

CompareExchange is confusing. Regardless of whether it makes a change or not it returns the old value in the ref location which is the first parameter. So you can't tell whether it's succeeded correctly by checking the return value alone in cases where multiple transitions from a single state are allowed, in this case we have moves from Unset to both Closed and Cancelled. So I've added checks for the value in the ref location after the CompareExchange.